atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampUtils.java

```diff
@@ -0,0 +1,244 @@
+     *
+     *  (a) a successful application of proposed changes, or
+     *  (b) an unsuccessful application, but the state of the world in Cassandra as indicated by the casResult
+     *      matches the casMap.
+     *
+     * Otherwise, it throws an IllegalStateException.
+     *
+     * @param casResult the CAS result to verify
+     * @param casMap the intended state of the world
+     * @throws IllegalStateException if the transaction was not successful, and the state of the world does not match
+     *         the intended state, as indicated in the casMap
+     */
+    public static void verifyCompatible(CqlResult casResult, Map<String, Pair<byte[], byte[]>> casMap) {
+        if (!isSuccessfullyApplied(casResult)) {
+            Set<Incongruency> incongruencies = getIncongruencies(casResult, casMap);
+            if (!incongruencies.isEmpty()) {
+                throw new IllegalStateException("[BACKUP/RESTORE] Observed incongruent state of the world; the"
+                        + " following incongruencies were observed: " + incongruencies);
+            }
+        }
+    }
+
+    private static Set<Incongruency> getIncongruencies(CqlResult casResult, Map<String, Pair<byte[], byte[]>> casMap) {
+        Map<String, byte[]> relevantCassandraState = getRelevantCassandraState(casResult, casMap);
+        return casMap.entrySet().stream()
+                .filter(entry ->
+                        !Arrays.equals(relevantCassandraState.get(entry.getKey()), entry.getValue().getRhSide()))
+                .map(entry -> createIncongruency(relevantCassandraState, entry.getKey(), entry.getValue().getRhSide()))
+                .collect(Collectors.toSet());
+    }
+
+    private static Map<String, byte[]> getRelevantCassandraState(
+            CqlResult casResult,
+            Map<String, Pair<byte[], byte[]>> casMap) {
+        return casResult.getRows().stream()
+                    .filter(row -> {
+                        String columnName = getColumnNameFromRow(row);
+                        return casMap.containsKey(columnName);
+                    })
+                    .collect(Collectors.toMap(
+                            CassandraTimestampUtils::getColumnNameFromRow,
+                            row -> getNamedColumnValue(row.getColumns(), VALUE_COLUMN)));
+    }
+
+    private static String getColumnNameFromRow(CqlRow row) {
+        return PtBytes.toString(getNamedColumnValue(row.getColumns(), COLUMN_NAME_COLUMN));
+    }
+
+    private static Incongruency createIncongruency(
+            Map<String, byte[]> relevantCassandraState,
+            String columnName,
+            byte[] desiredValue) {
+        return ImmutableIncongruency.builder()
+                .columnName(columnName)
+                .desiredState(desiredValue)
+                .actualState(relevantCassandraState.get(columnName))
+                .build();
+    }
+
+    private static boolean isSuccessfullyApplied(CqlResult casResult) {
```

Comment:
Should be moved up above `getIncongruencies`, so that `verifyCompatible` is easier to grok.